x/yarpctest/roundtrip_test.go

```diff
@@ -300,3 +303,77 @@ func TestServiceRouting(t *testing.T) {
 		})
 	}
 }
+
+func TestUnaryOutboundMiddleware(t *testing.T) {
+	p := NewPortProvider(t)
+
+	const (
+		service            = "service"
+		correctProcedure   = "correct-procedure"
+		incorrectProcedure = "inccorect"
+	)
+
+	mw := middleware.UnaryOutboundFunc(
+		func(ctx context.Context, req *transport.Request, next transport.UnaryOutbound) (*transport.Response, error) {
+			// fix procedure name
+			req.Procedure = correctProcedure
+			return next.Call(ctx, req)
+		})
+
+	tests := []struct {
+		name    string
+		service Lifecycle
+		request Action
+	}{
+		{
+			name: "HTTP",
+			service: HTTPService(
+				Name(service),
+				Proc(Name(correctProcedure), EchoHandler()),
+				p.NamedPort("http"),
+			),
+			request: HTTPRequest(
+				Service(service),
+				Procedure(incorrectProcedure),
+				UnaryOutboundMiddleware(mw),
+				p.NamedPort("http"),
+			),
+		},
+		{
+			name: "TChannel",
+			service: TChannelService(
+				Name(service),
+				Proc(Name(correctProcedure), EchoHandler()),
+				p.NamedPort("TChannel"),
+			),
+			request: TChannelRequest(
+				Service(service),
+				Procedure(incorrectProcedure),
+				UnaryOutboundMiddleware(mw),
+				p.NamedPort("TChannel"),
+			),
+		},
+		{
+			name: "gRPC",
+			service: GRPCService(
+				Name(service),
+				Proc(Name(correctProcedure), EchoHandler()),
+				p.NamedPort("grpc"),
+			),
+			request: GRPCRequest(
+				Service(service),
+				Procedure(incorrectProcedure),
+				UnaryOutboundMiddleware(mw),
+				p.NamedPort("grpc"),
+			),
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			require.NoError(t, tt.service.Start(t))
+			defer func() { require.NoError(t, tt.service.Stop(t)) }()
```

Comment:
let's make this an assert. defers are called when the goroutine exits. this includes when a panic happens or when `runtime.Goexit()` is called (which is the "kill the current goroutine" API). require.Foo relies on runtime.Goexit, so require inside a defer will end up calling `runtime.Goexit` even though we're already in that process.